server/services/oauth2/oauth2.go

```diff
@@ -0,0 +1,109 @@
+package oauth2
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/jellydator/ttlcache/v3"
+	"golang.org/x/oauth2"
+)
+
+type Config struct {
+	Issuer       string
+	ClientID     string
+	ClientSecret string
+	AuthURL      string
+	TokenURL     string
+}
+
+type Service struct {
+	config   *oauth2.Config
+	verifier *oidc.IDTokenVerifier
+
+	stateStorage        *ttlcache.Cache[string, string]
+	codeVerifierStorage *ttlcache.Cache[string, string]
+}
+
+func NewService(config Config, redirectURL string) (*Service, error) {
+	oauth2Config := &oauth2.Config{
+		ClientID:     config.ClientID,
+		ClientSecret: config.ClientSecret,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  config.AuthURL,
+			TokenURL: config.TokenURL,
+		},
+		RedirectURL: redirectURL,
+		Scopes:      []string{"openid", "profile"},
+	}
+
+	provider, err := oidc.NewProvider(context.Background(), config.Issuer)
+	if err != nil {
+		return nil, err
+	}
+	verifier := provider.Verifier(&oidc.Config{ClientID: config.ClientID})
+
+	stateStorage := ttlcache.New[string, string]()
+	codeVerifierStorage := ttlcache.New[string, string]()
+	go stateStorage.Start()
+	go codeVerifierStorage.Start()
```

Comment:
大丈夫だとは思うんだけど、このgoroutineがpanicするとアプリ全体が止まってしまうので、recoverを入れた方がいいのかもしれない。
```suggestion
	go func() {
    defer func() {
      if cause := recover(); cause != nil {
        log.Printf("panic") //TODO: いい感じのメッセージ
      }
    }()
    codeVerifierStorage.Start()
  }()
```

Reply:
panicするということはttlcacheが動いてないということなので、そのままpanicしてしまったほうがいい気がします。もしくは、recover時に再度Startさせるか。
今回はpanicすることもほとんどないと思うので、このままにしておきたいです。